lib/datadog/tracing/contrib/karafka/patcher.rb

```diff
@@ -24,24 +24,39 @@ def propagation
           # (e.g. `my_batch_operation messages.payloads`)
           # @see https://github.com/karafka/karafka/blob/b06d1f7c17818e1605f80c2bb573454a33376b40/README.md?plain=1#L29-L35
           def each(&block)
+            parent_span = Datadog::Tracing.active_span
+            parent_trace_digest = Datadog::Tracing.active_trace&.to_digest
+
             @messages_array.each do |message|
-              if configuration[:distributed_tracing]
-                headers = if message.metadata.respond_to?(:raw_headers)
-                  message.metadata.raw_headers
-                else
-                  message.metadata.headers
-                end
-                trace_digest = Karafka.extract(headers)
-                Datadog::Tracing.continue_trace!(trace_digest) if trace_digest
-              end
+              trace_digest = if configuration[:distributed_tracing]
+                               headers = if message.metadata.respond_to?(:raw_headers)
+                                           message.metadata.raw_headers
+                                         else
+                                           message.metadata.headers
+                                         end
+                               Karafka.extract(headers)
+                             end
 
-              Tracing.trace(Ext::SPAN_MESSAGE_CONSUME) do |span|
+              Tracing.trace(Ext::SPAN_MESSAGE_CONSUME, continue_from: trace_digest) do |span, trace|
                 span.set_tag(Ext::TAG_OFFSET, message.metadata.offset)
                 span.set_tag(Contrib::Ext::Messaging::TAG_DESTINATION, message.topic)
                 span.set_tag(Contrib::Ext::Messaging::TAG_SYSTEM, Ext::TAG_SYSTEM)
 
                 span.resource = message.topic
 
+                # link the outer trace (where the messages batch was consumed)
+                # with the individual message's processing trace, so they're easier to
+                # correlate in the Datadog UI
+                if parent_span && span.parent_id != parent_span.id
+                  # add a link from the parent trace to the message span
+                  span_link = Tracing::SpanLink.new(parent_trace_digest)
+                  span.links << span_link
+
+                  # add a link from the current trace to the parent span
+                  span_link = Tracing::SpanLink.new(trace.to_digest)
+                  parent_span.links << span_link
+                end
```

Comment:
I understand that _"span links"_ are not very widely used here in the ruby sdk, but one challenge we have once we _continue a trace_ is co-relating the span with its "original parent". I think _span links_ are a great fit for solving that problem.

Please let me know your thoughts!